File: src/main/services/mod-tools/mod-buffer-parser.test.ts
import assert from "node:assert/strict";

import type { Resource } from "@main/lib/mod-static-glb/types";
import { describe, it } from "vitest";

import { matchIndexResources } from "./mod-buffer-parser";

describe("matchIndexResources", () => {
    it("matches LOD index resources to hash-named position resources", () => {
        const positions = [
            {
                name: "789ae812Position",
                filename: "789ae812-Position.buf",
                stride: 40,
                values: {},
            },
            {
                name: "785b21f5Position",
                filename: "785b21f5-Position.buf",
                stride: 40,
                values: {},
            },
        ] satisfies Resource[];
        const indices = [
            {
                name: "_LOD0.789ae812_16590_0_Index",
                filename: "LOD0.789ae812-16590-0-Index.buf",
                format: "DXGI_FORMAT_R32_UINT",
                values: {},
            },
            {
                name: "_LOD0.785b21f5_57612_0_Index",
                filename: "LOD0.785b21f5-57612-0-Index.buf",
                format: "DXGI_FORMAT_R32_UINT",
                values: {},
            },
        ] satisfies Resource[];

        const matches = matchIndexResources(positions, indices, []);

        assert.deepEqual(
            matches.get("789ae812position")?.map((resource) => resource.name),
            ["_LOD0.789ae812_16590_0_Index"],
        );
        assert.deepEqual(
            matches.get("785b21f5position")?.map((resource) => resource.name),
            ["_LOD0.785b21f5_57612_0_Index"],
        );
    });

    it("matches index resources when named variants follow the buffer kind", () => {
        const positions = [
            {
                name: "ZhaoBodyPosition_Default",
                filename: ".\\1Default\\ZhaoBodyPosition.buf",
                stride: 40,
                values: {},
            },
            {
                name: "ZhaoBodyPosition_Bodysuit",
                filename: ".\\2Bodysuit\\ZhaoBodyPosition.buf",
                stride: 40,
                values: {},
            },
        ] satisfies Resource[];
        const indices = [
            {
                name: "ZhaoBodyAIB_Default",
                filename: ".\\1Default\\ZhaoBodyA.ib",
                format: "DXGI_FORMAT_R32_UINT",
                values: {},
            },
            {
                name: "ZhaoBodyAIB_Bodysuit",
                filename: ".\\2Bodysuit\\ZhaoBodyA.ib",
                format: "DXGI_FORMAT_R32_UINT",
                values: {},
            },
        ] satisfies Resource[];

        const matches = matchIndexResources(positions, indices, []);

        assert.deepEqual(
            matches.get("zhaobodyposition_default")?.map((resource) => resource.name),
            ["ZhaoBodyAIB_Default"],
        );
        assert.deepEqual(
            matches.get("zhaobodyposition_bodysuit")?.map((resource) => resource.name),
            ["ZhaoBodyAIB_Bodysuit"],
        );
    });
});

File: internal/tools/mod_buffer_parser_test.go
package tools

import "testing"

func TestMatchIndexResources(t *testing.T) {
	t.Parallel()

	t.Run("matches LOD index resources to hash-named position resources", func(t *testing.T) {
		t.Parallel()
		positions := []modBufferResource{
			{Name: "beef0021Position", Filename: "beef0021-Position.buf", Stride: 40},
			{Name: "beef0022Position", Filename: "beef0022-Position.buf", Stride: 40},
		}
		indices := []modBufferResource{
			{Name: "_LOD0.beef0021_16590_0_Index", Filename: "LOD0.beef0021-16590-0-Index.buf", Format: "DXGI_FORMAT_R32_UINT"},
			{Name: "_LOD0.beef0022_57612_0_Index", Filename: "LOD0.beef0022-57612-0-Index.buf", Format: "DXGI_FORMAT_R32_UINT"},
		}
		matches := matchIndexResources(positions, indices, nil)
		assertIndexNames(t, matches["beef0021position"], "_LOD0.beef0021_16590_0_Index")
		assertIndexNames(t, matches["beef0022position"], "_LOD0.beef0022_57612_0_Index")
	})

	t.Run("matches index resources when named variants follow the buffer kind", func(t *testing.T) {
		t.Parallel()
		positions := []modBufferResource{
			{Name: "ZhaoBodyPosition_Default", Filename: `.\1Default\ZhaoBodyPosition.buf`, Stride: 40},
			{Name: "ZhaoBodyPosition_Bodysuit", Filename: `.\2Bodysuit\ZhaoBodyPosition.buf`, Stride: 40},
		}
		indices := []modBufferResource{
			{Name: "ZhaoBodyAIB_Default", Filename: `.\1Default\ZhaoBodyA.ib`, Format: "DXGI_FORMAT_R32_UINT"},
			{Name: "ZhaoBodyAIB_Bodysuit", Filename: `.\2Bodysuit\ZhaoBodyA.ib`, Format: "DXGI_FORMAT_R32_UINT"},
		}
		matches := matchIndexResources(positions, indices, nil)
		assertIndexNames(t, matches["zhaobodyposition_default"], "ZhaoBodyAIB_Default")
		assertIndexNames(t, matches["zhaobodyposition_bodysuit"], "ZhaoBodyAIB_Bodysuit")
	})

	t.Run("does not treat an earlier ib inside the base name as the index kind", func(t *testing.T) {
		t.Parallel()
		positions := []modBufferResource{
			{Name: "Rib_BodyPosition", Filename: "Rib_BodyPosition.buf", Stride: 40},
		}
		indices := []modBufferResource{
			{Name: "Rib_Body_Index", Filename: "Rib_Body_Index.buf", Format: "DXGI_FORMAT_R32_UINT"},
		}
		matches := matchIndexResources(positions, indices, nil)
		assertIndexNames(t, matches["rib_bodyposition"], "Rib_Body_Index")
	})
}

func assertIndexNames(t *testing.T, resources []modBufferResource, want ...string) {
	t.Helper()
	if len(resources) != len(want) {
		t.Fatalf("matches = %#v, want %v", resources, want)
	}
	for i, name := range want {
		if resources[i].Name != name {
			t.Fatalf("matches[%d] = %q, want %q", i, resources[i].Name, name)
		}
	}
}
